### PR TITLE
Add StartProductionAI to ProductionQueue to fix duplicated production on AI module caused by order latency

### PR DESCRIPTION
**Pull request by dnqbob**
This PR fix the duplicated production due to order latency (or other methods to handle "network lags")

## How to test
Testcase (Fixed)
1. Runing this PR twice, you get two game program(A and B).
2. Enable AI debug message
3. A create a room in mutiplayer game, B join it.
4. Get at least one AI player in game, and start the game
Result: you will find AI module always produce duplicatedly in debug message, but it has no effect on AI gameplay.

Testcase (Unfixed)
1. Revert the fixing commit, and do 1~5 steps in "Testcase (Fixed)"
Result: you will find AI module always produce duplicatedly.

#### Calculate a delay from orderlantency to override the `StructureProductionActiveDelay`
This idea has following problem:

1. The calculation itself is very difficult.
The best and adaptive result I get is `Ticks = OrderLatency * 10 / 3` to testcase, but Idk if it is the correct answer for all.

2. Calculated delay is always too large for singleplayer gameplay. 
Singleplayer mode actually don't have order latency and if you apply calculated delay AI will looks slow and stupid (delay > 50 ticks). Unless we expose detail of singleplayer state to AI modules.

3. No longer works in the future plan.
If the order latency setting becomes dynamic or dropped in the future (see "Server.cs", line 1407),  this idea will no longer work because the delay will be unpredicable.

#### Allow user set up mutiple `StructureProductionActiveDelay` for each game speed
This idea has following problem:
1. User will have a difficult time to set up those delays, and hard to maintain.
Each delay needs testing from local to server, and if order latency is changed user have to do this again.

2. Singleplayer gameplay needs another delay. 
Singleplayer mode actually don't have orderlatency, and we need to expose singleplayer state to AI modules.

3. No longer works in the future plan.
If the order latency setting becomes dynamic or dropped in the future (see "Server.cs", line 1407),  this idea will no longer work because the delay will be unpredicable.

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs [hunk 1]
```diff
@@ -392,85 +392,93 @@ public bool CanQueue(ActorInfo actor, out string notificationAudio, out string n
 			return true;
 		}
 
-		public void ResolveOrder(Actor self, Order order)
+		void StartProduction(Actor self, Order order)
 		{
-			if (!Enabled)
+			var rules = self.World.Map.Rules; var unit = rules.Actors[order.TargetString];
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs]:
```suggestion
			var rules = self.World.Map.Rules;
			var unit = rules.Actors[order.TargetString];
```

File: OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs [hunk 2]
```diff
@@ -41,7 +41,7 @@ public class UnitBuilderBotModuleInfo : ConditionalTraitInfo
 
 	public class UnitBuilderBotModule : ConditionalTrait<UnitBuilderBotModuleInfo>, IBotTick, IBotNotifyIdleBaseUnits, IBotRequestUnitProduction, IGameSaveTraitData
 	{
-		public const int FeedbackTime = 30; // ticks; = a bit over 1s. must be >= netlag.
+		public const int FeedbackTime = 30;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs:47]: I think its ok/good to mention these are ticks and not ms.

```suggestion
		/// <summary>
		/// Feedback time in ticks.
		/// </summary>
		public const int FeedbackTime = 30;
```

File: OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs]: fixes

File: OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs (hunk 2 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs:47]: fixed

File: OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs [hunk 3]
```diff
@@ -392,85 +392,94 @@ public bool CanQueue(ActorInfo actor, out string notificationAudio, out string n
...
+
+			switch (order.OrderString)
+			{
+				case "StartProduction":
+					StartProduction(self, order);
+					break;
+				case "StartProductionAI":
+					// Hack: AI production modules won't produce on queue that is producing
+					// to avoid duplicated production due to order latency or network lags.
+					if (!AllQueued().Any())
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Traits/Player/ProductionQueue.cs:478]: Why don't you move this into `UnitBuilderBotModule` and avoid the AI related code here with the strange custom order string?

**Comment by Mailaender** [comment]
Does this resolve https://github.com/OpenRA/OpenRA/issues/14024?

**Comment by dnqbob** [comment]
> Does this resolve #14024?

~~Yes, it can solve #14024.~~
No, there is a flaw in my thought. This PR can fix the problem on singlequeue production but not for mutiqueue production when there is more than 1 conyards.

**Comment by dnqbob** [comment]
It seems the only idea we have is check building in all production queue and cancel the exceed one waiting in line or constructing.

**Pull request closed without merging**